src/sentry/static/sentry/app/components/lazyLoad.jsx

```diff
@@ -1,35 +1,119 @@
-import React from 'react';
 import PropTypes from 'prop-types';
+import Raven from 'raven-js';
+import React from 'react';
 
+import {t} from '../locale';
+import LoadingError from './loadingError';
 import LoadingIndicator from '../components/loadingIndicator';
 
 class LazyLoad extends React.Component {
   static propTypes = {
     hideBusy: PropTypes.bool,
+    hideError: PropTypes.bool,
     /**
-     * specifically needs to be a thenable
+     * Function that returns a promise of a React.Component
+     * Alias property for `getComponent` to maintain compatibility
      */
     component: PropTypes.func,
+
+    /**
+     * Function that returns a promise of a React.Component
+     */
+    getComponent: PropTypes.func,
+
+    /**
+     * Also accepts a route object from react-router that has a `getComponent` property
+     */
+    route: PropTypes.shape({
+      getComponent: PropTypes.func,
+    }),
   };
 
   constructor(...args) {
     super(...args);
     this.state = {
       Component: null,
+      error: null,
     };
   }
 
   componentDidMount() {
-    this.props.component().then(Component => {
-      this.setState({Component});
+    this.fetchComponent();
+  }
+
+  componentWillReceiveProps(nextProps, nextState) {
+    // This is to handle the following case:
+    // <Route path="a/">
+    //   <Route path="b/" component={LazyLoad} getComponent={...} />
+    //   <Route path="c/" component={LazyLoad} getComponent={...} />
+    // </Route>
+    //
+    // `LazyLoad` will get not fully remount when we switch between `b` and `c`,
+    // instead will just re-render.  Refetch if locations are different
+    if (nextProps.location.pathname === this.props.location.pathname) return;
+
+    this.setState({
+      Component: null,
     });
+
+    this.fetchComponent();
   }
 
+  getComponentGetter = () =>
+    this.props.component || this.props.getComponent || this.props.route.getComponent;
+
+  fetchComponent = () => {
+    let getComponent = this.getComponentGetter();
+
+    getComponent()
+      .then(
+        Component => {
+          this.setState({
+            Component,
+          });
+        },
+        err => {
+          this.setState({
+            error: err,
+          });
+        }
+      )
+      .catch(err => {
+        // eslint-disable-next-line no-console
+        console.warn(err);
+        Raven.captureException(err);
```

Comment:
maybe there should be an option for hideerror that doesn't send them to sentry?

Reply:
when would we not want to capture errors? I think we can address in a future pr when needed too